File: src/main/java/com/hileco/cortex/instructions/io/IoInstruction.java
package com.hileco.cortex.instructions.io;


import com.hileco.cortex.vm.ProgramZone;
import com.hileco.cortex.vm.ProgramStoreZone;
import com.hileco.cortex.instructions.Instruction;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;

import java.util.List;

import static com.hileco.cortex.vm.ProgramZone.CALL_DATA;
import static com.hileco.cortex.vm.ProgramZone.DISK;
import static com.hileco.cortex.vm.ProgramZone.MEMORY;
import static com.hileco.cortex.vm.ProgramZone.STACK;

@Getter
@AllArgsConstructor
@EqualsAndHashCode
public abstract class IoInstruction implements Instruction {
    private final ProgramStoreZone programStoreZone;

    private static List<ProgramZone> programZoneFor(ProgramStoreZone programStoreZone) {
        switch (programStoreZone) {
            case MEMORY:
                return List.of(STACK, MEMORY);
            case DISK:
                return List.of(STACK, DISK);
            case CALL_DATA:
                return List.of(STACK, CALL_DATA);
            default:
                throw new IllegalArgumentException(String.format("Unsupported ProgramStoreZone: %s", programStoreZone));
        }
    }

    @Override
    public List<ProgramZone> getInstructionModifiers() {
        return programZoneFor(this.programStoreZone);
    }

    @Override
    public String toString() {
        return String.format("%s %s", this.getClass().getSimpleName(), this.programStoreZone);
    }
}

File: src/main/java/com/hileco/cortex/instructions/io/IoInstruction.kt
package com.hileco.cortex.instructions.io


import com.hileco.cortex.instructions.Instruction
import com.hileco.cortex.vm.concrete.ProgramStoreZone
import com.hileco.cortex.vm.concrete.ProgramZone
import com.hileco.cortex.vm.concrete.ProgramZone.*

abstract class IoInstruction(val programStoreZone: ProgramStoreZone) : Instruction() {
    override val instructionModifiers: List<ProgramZone>
        get() = when (programStoreZone) {
            ProgramStoreZone.MEMORY -> listOf(STACK, MEMORY)
            ProgramStoreZone.DISK -> listOf(STACK, DISK)
            ProgramStoreZone.CALL_DATA -> listOf(STACK, CALL_DATA)
        }

    override fun toString(): String {
        return "${this::class.simpleName} $programStoreZone"
    }
}
